File: Linked List/Singly Linked List/LinkedList.py
# Program to perform various functionalities on singly linked list

# class to create a node
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


# class to provide different functionalities of linked list
class LinkedList:
    def __init__(self):
        self.head = None

    def insert_from_tail(self, data):
        node = Node(data)
        if self.head:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = node
        else:
            self.head = node

    def insert_from_head(self, data):
        node = Node(data)
        if self.head:
            temp = node
            temp.next = self.head
            self.head = temp
        else:
            self.head = node

    def delete_from_head(self):
        if self.head:
            temp = self.head
            print(f"\nDeleted Element: {temp.data}")
            self.head = temp.next
            del temp
        else:
            print("Linked list is empty!!")

    def delete_from_tail(self):
        if self.head:
            temp, temp_1 = self.head, self.head
            while temp.next:
                temp = temp.next
            while temp_1.next != temp:
                temp_1 = temp_1.next
            print(f"\nDeleted Element: {temp.data}")
            temp_1.next = None
            del temp
        else:
            print("Linked list is empty!!")

    def display(self):
        if self.head:
            print("Elements of the linked list:", end=" ")
            current = self.head
            while current:
                print(current.data, end="->")
                current = current.next
            print(None)
        else:
            print("Linked list is empty!!")


# driver
if __name__ == '__main__':
    linked_list = LinkedList()

    choice = 1
    while choice != 6:
        print("1. Insert the element from head")
        print("2. Insert the element from tail")
        print("3. Delete the element from head")
        print("4. Delete the element from tail")
        print("5. Display the linked list")
        print("6. Exit")

        choice = int(input("Enter your choice: "))
        match choice:
            case 1:
                element = int(input("Enter the element: "))
                linked_list.insert_from_head(element)
            case 2:
                element = int(input("Enter the element: "))
                linked_list.insert_from_tail(element)
            case 3:
                linked_list.delete_from_head()
            case 4:
                linked_list.delete_from_tail()
            case 5:
                linked_list.display()
            case 6:
                print("!! Thanks for using the console !!")
                exit()
            case _:
                print("Kindly select right choice!!")

File: Linked List/Singly Linked List/LinkedList.cpp
// Program to implement Linked List

#include<iostream>

using namespace std;

class node
{
    public:
    int data;
    node *next;

    node(int item)
    {
        data = item;
        next = NULL;
    }
};

class LinkedList
{
    node *head;

    public:
    LinkedList()
    {
        head = NULL;
    }

    // Function to create a Linked List
    void createLinkedList(int item)
    {
        node *newnode = new node(item);
        node *temp = head;

        if(head == NULL)
        {
            head = temp = newnode;
            newnode -> next = NULL;
        }
        else
        {
            while(temp -> next != NULL)
            {
                temp = temp -> next;
            }
            temp -> next = newnode;
            newnode -> next = NULL;
        }
    }

    // Function to insert node at first position
    void insertNodeAtFirst(int item)
    {
        node *newnode = new node(item);
        node *temp = head;

        if(head == NULL)
        {
            head = temp = newnode;
            newnode -> next = NULL;
        }
        else
        {
            temp = newnode;
            newnode -> next = head;
            head = newnode;
        }
    }

    // Function to insert node at last position
    void insertNodeAtLast(int item)
    {
        node *newnode = new node(item);
        node *temp = head;

        if(head == NULL)
        {
            head = temp = newnode;
            newnode -> next = NULL;
        }
        else
        {
            while(temp -> next != NULL)
            {
                temp = temp -> next;
            }
            temp -> next = newnode;
            newnode -> next = NULL;
        }
    }

    // Function to insert node after the key inserted by user
    void insertNodeAfterKey(int key, int item)
    {
        node *newnode = new node(item);
        node *temp = head;

        while(temp != NULL)
        {
            if(temp -> data == key)
            {
                newnode -> next = temp -> next;
                temp -> next = newnode;
            }

            temp = temp -> next;
        }
    }

    // Function to delete the first node
    void deleteAtFirst()
    {
        node *temp = head;

        if(head == NULL)
        {
            cout << "\nThe Linked List is empty !!" << endl;
        }
        else
        {
            cout << "\nThe deleted element: " << head -> data << endl;
            head = head -> next;
            delete(temp);
        }
    }

    // Function to delete the last node
    void deleteAtLast()
    {
        node *temp_1 = head;
        node *temp = head;

        if(head == NULL)
        {
            cout << "\nThe Linked List is empty !!" << endl;
        }
        else
        {
            while(temp -> next != NULL)
            {
                temp = temp -> next; 
            }
            while(temp_1 -> next != temp)
            {
                temp_1 = temp_1 -> next;
            }
            cout << "\nThe deleted element: " << temp -> data << endl;
            temp_1 -> next = NULL;
            delete(temp);
        }
    }

    //Function to delete the key entered by user
    void deleteNodeAtKey(int key)
    {
        node *temp = head;
        node *previous = NULL;

        while(temp != NULL)
        {
            if(temp -> data == key)
            {
                if(temp == head)
                {
                    head = head -> next;
                    cout << "\nThe deleted element: " << temp -> data << endl;
                    delete(temp);
                    temp = head;
                }
                else
                {
                    previous -> next = temp -> next;
                    cout << "\nThe deleted element: " << temp -> data << endl;
                    delete(temp);
                    temp = previous -> next;
                }
            }
            else
            {
                previous = temp;
                temp = temp -> next;
            }
        }
    }

    // Function to display the linked list
    void display()
    {
        node *temp = head;

        if(head == NULL)
        {
            cout << "\nThe Linked List is empty !!" << endl;
        }
        else
        {
            cout << "\nLinked List" << endl;
            while(temp != NULL)
            {
                cout << temp -> data << " ";
                temp = temp -> next;
            }
            cout << endl;
        }
    }
};

int main()
{
    LinkedList L;
    int item, key;
    int choice = 1;
    
    while(choice != 9)
    {
        cout << "\n*** OPERATIONS ***" << endl;
        cout << "1.  Create a single linked list" << endl;
        cout << "2.  Insert node at first position" << endl;
        cout << "3.  Insert node at last position" << endl;
        cout << "4.  Insert a node after a given key value" << endl;
        cout << "5.  Delete node at first position" << endl;
        cout << "6.  Delete node at last position" << endl;
        cout << "7.  Delete a node containing given key value" << endl;
        cout << "8.  Display the linked list" << endl;
        cout << "9.  Exit" << endl;

        // Scanning the choice from the user
        cout << "\nEnter your choice: ";
        cin >> choice;

        switch(choice)
        {
            case 1:
            {
                // Scannig the element to enter in Linked List
                cout << "Enter the value: ";
                cin >> item;

                // Calling the createLinkedList function
                L.createLinkedList(item);
                break;
            }

            case 2:
            {
                // Scannig the element to enter in Linked List
                cout << "Enter the value: ";
                cin >> item;

                // Calling the insertNodeAtFirst function
                L.insertNodeAtFirst(item);
                break;
            }

            case 3:
            {
                // Scannig the element to enter in Linked List
                cout << "Enter the value: ";
                cin >> item;

                // Calling the insertNodeAtLast function
                L.insertNodeAtLast(item);
                break;
            }

            case 4:
            {
                // Scannig the element to enter in Linked List
                cout << "Enter the value: ";
                cin >> item;

                // Scanning the key element after which the item would insert
                cout << "Enter the element after which you want to add the node: ";
                cin >> key;

                // Calling insertNodeAfterKey function
                L.insertNodeAfterKey(key,item);
                break;
            }

            case 5:
            {
                // Calling deleteAtFirst function
                L.deleteAtFirst();
                break;
            }

            case 6:
            {
                // Calling deleteAtLast function
                L.deleteAtLast();
                break;
            }

            case 7:
            {
                // Scanning the element to delete
                cout << "Enter the element you want to delete: ";
                cin >> key;

                // Calling deleteNodeAtKey function
                L.deleteNodeAtKey(key);
                break;
            }

            case 8:
            {
                // Calling display function
                L.display();
                break;
            }

            default:
            {
                if(choice > 9)
                {
                    cout << "\nEnter Right Choice !!" << endl;
                }
            }
        }
    }

    return 0;
}
